clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3007,19 +3007,37 @@ class SYCLKernelNameTypeVisitor
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
     const DeclContext *DeclCtx = Tag->getDeclContext();
     if (DeclCtx && !UnnamedLambdaEnabled) {
-      auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
-      if (NameSpace && NameSpace->isStdNamespace()) {
-        S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << KernelNameType;
-        S.Diag(KernelInvocationFuncLoc, diag::note_invalid_type_in_sycl_kernel)
-            << /* kernel name cannot be a type in the std namespace */ 2
-            << QualType(Tag->getTypeForDecl(), 0);
-        IsInvalid = true;
-        return;
+      // Loop through Declaration Contexts to detect kernel names nested inside
+      // "std" and "anonymous" namespaces.
+      while (!DeclCtx->isTranslationUnit()) {
+        auto *NSDecl = dyn_cast<NamespaceDecl>(DeclCtx);
+        if (NSDecl && NSDecl->isStdNamespace()) {
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::err_sycl_kernel_incorrectly_named)
+              << KernelNameType;
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::note_invalid_type_in_sycl_kernel)
+              << /* kernel name cannot be a type in the std namespace */ 2
+              << QualType(Tag->getTypeForDecl(), 0);
+          IsInvalid = true;
+          return;
+        }
+        if (NSDecl && NSDecl->isAnonymousNamespace()) {
```

Comment:
I think this code should be combined with the above code, along the lines of:
```
if (NSDecl) {
  bool IsStd = NSDecl->isStdNamespace();
  if (IsStd || NSDecl->isAnonymousNamespace()) {
    S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
      << KernelNameType << (IsStd ? 2 : 0) << Tag;
    IsInvalid = true;
    return;
  }
}
```
(This presumes the note was combined with the diagnostic.)

Reply:
I can do this once we get consensus on whether we need both the error and note diagnostics.

Reply:
I won't be able to combine the code as above, since 'std' ns prints both kernel name and type, where as we just print the kernel name for anonymous namespace
Example:
For 'std' ns:
`'std::NestedInStd::NestedStruct' is an invalid kernel name, type 'std::NestedInStd::NestedStruct' cannot be in the 'std' namespace`
For anonymous ns:
`'(anonymous namespace)::NestedInAnon::StructInAnonymousNS' should be globally-visible`